### PR TITLE
Make `knn` search requests fully cancellable

### LINKED ISSUE #90323: Make sure ANN searches are cancellable
Currently if you provide an ANN search using the `knn` search section, the search doesn't respect cancellation. This is because ANN searches use their own Lucene data structures that aren't covered by `ExitableDirectoryReader`.

This class is forked from Lucene, and in Lucene's version we recently added fine-grained cancellation for ANN search. So I think we just need to fold in the changes to the Elasticsearch version.

### PR DESCRIPTION
**Pull request by benwtrent**
Approximate nearest-neighbor search requests should be cancelled if the overall search request is cancelled. Currently, the cancellation is not respected by searches provided in the `knn` section of the search request.

This commit ports over the `ExitableDirectoryReader` changes made in Lucene. The main differences (besides formatting) is changing the timeout checker to be our query cancellation checker.

Related to: https://github.com/apache/lucene/pull/833

closes: https://github.com/elastic/elasticsearch/issues/90323

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java [hunk 1]
```diff
@@ -12,14 +12,18 @@
...
 import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.index.PointValues;
 import org.apache.lucene.index.QueryTimeout;
 import org.apache.lucene.index.Terms;
 import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.index.VectorValues;
 import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.TopDocs;
 import org.apache.lucene.search.suggest.document.CompletionTerms;
+import org.apache.lucene.util.Bits;
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/search/internal/ExitableDirectoryReader.java:26]: This is not related to your change, but I noticed that the Javadoc on this class doesn't explain it's been copied and adapted from Lucene. You can see `KnnScoreDocQuery` for an example comment explaining how / why it was copied.

**Review by jtibshirani — COMMENTED** [review]
Thanks @benwtrent for the speedy fix! We should also make sure this addition is tested. One option is to add a new case to `SearchCancellationTests`. It seems even better if we made sure to copy over and adapt `TestExitableDirectoryReader` from Lucene (hmm, maybe we could have done this earlier!)

**Comment by benwtrent** [comment]
@jtibshirani I was wondering about tests. I was looking around to see if we had any cancellable search tests and... well, there ain't many. 

I will see about adding one. 

Copying `TestExitableDirectoryReader` Seems prudent as well.

**Comment by benwtrent** [comment]
@jtibshirani I added a test case for `SearchCancellationTests`. It seems that this set of tests gives us coverage that is comparable to `TestExitableDirectoryReader`. We could as a future refactor remove `SearchCancellationTests` and make it closer akin to `TestExitableDirectoryReader`

**Review by jtibshirani — APPROVED** [review]
This looks good to me! I agree with the version labelling too -- it's not clearly a "bugfix" and I don't see a strong reason to try to backport it to 8.5.

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**